Adjust Onboarding's ContinueButton layout based on Mobile/Desktop target

johnny9 opened the pull request:
This change registers a new QObject that can be used to check if the app is running in DESKTOP or MOBILE mode. Based on that value, adjustments to the ContinueButton's anchors will be made to match the spec for each configuration. 

[![Windows](https://img.shields.io/badge/OS-Windows-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/win64/insecure_win_gui.zip?branch=pull/171)
[![Intel macOS](https://img.shields.io/badge/OS-Intel%20macOS-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos/insecure_mac_gui.zip?branch=pull/171)
[![Apple Silicon macOS](https://img.shields.io/badge/OS-Apple%20Silicon%20macOS-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/macos_arm64/insecure_mac_arm64_gui.zip?branch=pull/171)
[![ARM64 Android](https://img.shields.io/badge/OS-Android-green)](https://api.cirrus-ci.com/v1/artifact/github/bitcoin-core/gui-qml/android/insecure_android_apk.zip?branch=pull/171)

johnny9 commented:
After looking through this a bit, I believe the mobile and desktop specs are close enough that most of the top level layouts can be reused without having to provide a different root QML for mobile.  Certain components, however, will end up with more complex anchors/layout properties due to having two different states they can be in.

**DESKTOP:**
![onboarding_1](https://user-images.githubusercontent.com/985648/190837086-5432af9f-5325-46e5-8adf-55db891b0133.png)

**MOBILE:**
![Screenshot_20220916-212819](https://user-images.githubusercontent.com/985648/190837264-826186bf-b163-47e4-bfe7-deabad04ba4e.png)

jarolrod commented:
Concept ACK on fixing up the UI to fit our mobile design

jarolrod commented on src/qml/bitcoin.cpp:
```diff
@@ -13,6 +13,7 @@
 #include <noui.h>
 #include <qml/imageprovider.h>
 #include <qml/nodemodel.h>
+#include <qml/appmode.h>
```
For 872807a05daeae9752a39df275a3aff602c942e4, please make the following style changes:

``` diff
diff --git a/src/qml/appmode.h b/src/qml/appmode.h
index 0148a4240..e45977318 100644
--- a/src/qml/appmode.h
+++ b/src/qml/appmode.h
@@ -15,14 +15,12 @@ class AppMode : public QObject
     Q_PROPERTY(QString state READ state NOTIFY modeChanged)
 
 public:
-    enum Mode
-    {
+    enum Mode {
         DESKTOP,
         MOBILE
     };
 
-    explicit AppMode(Mode mode) :
-        m_mode(mode)
+    explicit AppMode(Mode mode) : m_mode(mode)
     {
     }
 
@@ -31,12 +29,12 @@ public:
     QString state()
     {
         switch (m_mode) {
-            case MOBILE:
-                return "MOBILE";
-            case DESKTOP:
-                return "DESKTOP";
-            default:
-                return "DESKTOP";
+        case MOBILE:
+            return "MOBILE";
+        case DESKTOP:
+            return "DESKTOP";
+        default:
+            return "DESKTOP";
         }
     }
     Mode mode() const { return m_mode; }
diff --git a/src/qml/bitcoin.cpp b/src/qml/bitcoin.cpp
index c8687b5ed..291b39f53 100644
--- a/src/qml/bitcoin.cpp
+++ b/src/qml/bitcoin.cpp
@@ -11,9 +11,9 @@
 #include <logging.h>
 #include <node/interface_ui.h>
 #include <noui.h>
+#include <qml/appmode.h>
 #include <qml/imageprovider.h>
 #include <qml/nodemodel.h>
-#include <qml/appmode.h>
 #include <qml/util.h>
 #include <qt/guiconstants.h>
 #include <qt/guiutil.h>
```

You can get the same diff locally by running the following command on 872807a05daeae9752a39df275a3aff602c942e4
```
git diff -U0 HEAD~1.. | ./contrib/devtools/clang-format-diff.py -p1 -i -v
```

johnny9 replied:
Thanks, that's really good info on the cpp style too.

jarolrod commented on src/Makefile.qt.include:
```diff
@@ -110,6 +111,7 @@ BITCOIN_QT_H = \
   qml/bitcoin.h \
   qml/imageprovider.h \
   qml/nodemodel.h \
+  qml/appmode.h \
```
please keep the list in alphabetical order, so we would make the following change to the file:

``` diff
diff --git a/src/Makefile.qt.include b/src/Makefile.qt.include
index 096cd116e..06b642d93 100644
--- a/src/Makefile.qt.include
+++ b/src/Makefile.qt.include
@@ -35,8 +35,8 @@ QT_FORMS_UI = \
   qt/forms/transactiondescdialog.ui
 
 QT_MOC_CPP = \
-  qml/moc_nodemodel.cpp \
   qml/moc_appmode.cpp \
+  qml/moc_nodemodel.cpp \
   qt/moc_addressbookpage.cpp \
   qt/moc_addresstablemodel.cpp \
   qt/moc_askpassphrasedialog.cpp \
@@ -108,10 +108,10 @@ QT_QRC_LOCALE_CPP = qt/qrc_bitcoin_locale.cpp
 QT_QRC_LOCALE = qt/bitcoin_locale.qrc
 
 BITCOIN_QT_H = \
+  qml/appmode.h \
   qml/bitcoin.h \
   qml/imageprovider.h \
   qml/nodemodel.h \
-  qml/appmode.h \
   qml/util.h \
   qt/addressbookpage.h \
   qt/addresstablemodel.h \
```

jarolrod reviewed (commented):
will continue testing, but some style fixups for now

johnny9 commented:
Updated from  6412a3a to  1d10e8e

Changes:
 - fixed style issues in appmode.h
 - fixed include oridering in Makefile.qt.include to alphabetical
 - removed width changes to onboarding01-03 that are meant for a future commit

johnny9 commented on src/qml/bitcoin.cpp:
```diff
@@ -184,6 +185,14 @@ int QmlGuiMain(int argc, char* argv[])
+    AppMode app_mode(AppMode::DESKTOP);
+#endif // __ANDROID__
+
+    qmlRegisterSingletonInstance<AppMode>("com.bitcoincore.qt", 1, 0, "AppMode", &app_mode);
```
not 100% sure that is the namespace that should be used but it is what was defined in the AndroidManifest.xml also.  org.bitcoin.qt might make more sense for both.

jarolrod replied:
@hebasto what do you think about this URI namespace, see: https://doc.qt.io/qt-6/qqmlengine.html#qmlRegisterSingletonType

jarolrod commented:
Confirming this does as the description states

This pr on an android tablet:

| | | |
| - | - | - |
| ![Screenshot_20220920-210432](https://user-images.githubusercontent.com/23396902/191403492-e02ef2cb-df19-45b1-aa86-c38ecd333c34.jpg) | ![Screenshot_20220920-210523](https://user-images.githubusercontent.com/23396902/191403614-a1caa2d8-4067-4d9b-9278-87c838d2bfaf.jpg) | ![Screenshot_20220920-210532](https://user-images.githubusercontent.com/23396902/191403629-a5241041-39f6-4b5e-ace1-ae2555789dad.jpg) |

This pr on desktop:

| | | |
| - | - | - |
| <img width="752" alt="Screen Shot 2022-09-20 at 9 55 57 PM" src="https://user-images.githubusercontent.com/23396902/191403738-f766ab5b-c79d-49a2-af07-a8747bbf3544.png"> | <img width="752" alt="Screen Shot 2022-09-20 at 9 56 03 PM" src="https://user-images.githubusercontent.com/23396902/191403775-e90b3f85-d2d8-4238-9938-d1480c0b515f.png"> | <img width="752" alt="Screen Shot 2022-09-20 at 9 56 07 PM" src="https://user-images.githubusercontent.com/23396902/191403797-877beecb-ac3f-4552-8bc5-1db91b58a399.png"> |

johnny9 commented:
Updated  from 1d10e8e to eaba2d4 
Changes:
  - added newline to end of appmode.h
  - squashed commits

johnny9 commented:
Updated from[ eaba2d4 ](https://github.com/bitcoin-core/gui-qml/commit/eaba2d4f6e52c690ab37193b1ede3173129c5274)to[ cd7f5a8](https://github.com/bitcoin-core/gui-qml/commit/cd7f5a837887366175be7ce74586bdb0fe46b6bf): 

  - Rebased with main

johnny9 commented:
Updated from[ cd7f5a8 ](https://github.com/bitcoin-core/gui-qml/commit/cd7f5a837887366175be7ce74586bdb0fe46b6bf)to 8a3ee65

Changes:
  - Set the qml module namespace to org.bitcoincore.qt

jarolrod commented on src/qml/controls/OnboardingInfo.qml:
```diff
@@ -49,11 +53,40 @@ Control {
+
+    state: AppMode.state
+
+    states: [
```
An alternative would be to define anchor values for components inside of our theme file, and assign values in the same way we assign the color values; but when the mode changes, the anchor values change. I’ll make a branch implementing this idea and we can compare

jarolrod replied:
The gist of the idea is: 

This checking of the AppMode state can just be hidden inside of the Theme file, and the theme passes the values we want to the component

johnny9 replied:
I'm not sure if it makes sense to move anchor values outside of the component definition since they are relative values that only make sense in a context that includes all of the components involved.

One thing I also learned while doing this is that Qt doesn't recommend conditional anchors bindings. "Because the evaluation order of bindings is not defined, it is not recommended to change anchors via conditional bindings, as this can lead to the ordering issue described above. In the following example the Rectangle will eventually grow to the full width of its parent, because both left and right anchors will be simultaneously set during binding update." (https://doc.qt.io/qt-6/qtquick-positioning-anchors.html#changing-anchors). That was my primary reason for using states and AnchorChanges which unfortunately requires more boilerplate than just using a tertiary.

jarolrod reviewed (commented):
Strong Concept ACK of course, just one thing that may be worth considering

hebasto commented on src/qml/controls/OnboardingInfo.qml:
```diff
@@ -5,8 +5,9 @@
 import QtQuick 2.15
 import QtQuick.Controls 2.15
 import QtQuick.Layouts 1.15
+import com.bitcoincore.qt 1.0
```
It differs from `qmlRegisterSingletonInstance` call argument.

hebasto reviewed (commented):
Great to see such a PR that introduces mobile-specific code!

> This change registers a new QObject that can be used to check if the app is running in DESKTOP or MOBILE mode. 

At some point in the past we thought that desktop/mobile code will reside in separated directories. Using dedicated states for different platforms also looks ok.

Should we distinguish Android and iOS mobile devices?

johnny9 commented:
Update from[ 8a3ee65 ](https://github.com/bitcoin-core/gui-qml/commit/8a3ee6581ece59e8fb23debd7c8370f8d9a70617)to 038caba:
  - Rebased with main
  - Fix a bad import from "com.bitcoincore.qt" to "org.bitcoincore.qt"

jarolrod reviewed (approved):
ACK 038caba43dbdda4b6e95a1da08272559c9623a9f

I have reviewed the code and agree that this can be merged. Also tested on both desktop and Android 12:

### Desktop
#### master

|    |    |
| -- | -- |
| <img width="752" alt="Screen Shot 2022-10-03 at 11 37 42 PM" src="https://user-images.githubusercontent.com/23396902/193736420-12256b7a-90c8-4800-a85d-cf78ad66aae5.png"> | <img width="752" alt="Screen Shot 2022-10-03 at 11 37 47 PM" src="https://user-images.githubusercontent.com/23396902/193736434-bbc1b977-9388-4e19-90c8-65d06ca0adda.png"> |

#### PR
|    |    |
| -- | -- |
| <img width="708" alt="Screen Shot 2022-10-03 at 11 30 01 PM" src="https://user-images.githubusercontent.com/23396902/193736538-5909a0ee-fb44-41e9-a6ec-5430464146ab.png"> | <img width="752" alt="Screen Shot 2022-10-03 at 11 30 14 PM" src="https://user-images.githubusercontent.com/23396902/193736563-47dd9bad-0daf-4f1f-b935-a6aab084d8d7.png"> |

### Mobile

#### master
|    |    |
| -- | -- |
| ![0-1](https://user-images.githubusercontent.com/23396902/193736758-a8df11c7-2fbb-4ddf-a2d8-7d812321b8fc.jpeg) | ![0](https://user-images.githubusercontent.com/23396902/193736773-1ec4ec17-860c-44f4-958a-491f1d3fbd68.jpeg) |



#### PR
|    |    |
| -- | -- |
| ![0-1](https://user-images.githubusercontent.com/23396902/193736808-72148d3a-9494-49c7-8d1b-9d974500bc10.jpeg) | ![0](https://user-images.githubusercontent.com/23396902/193736825-e5cb5289-5831-47f9-8e8d-2c47f93f0ad7.jpeg) |
